ambry-router/src/main/java/com/github/ambry/router/CompressionService.java

```diff
@@ -250,32 +249,25 @@ public ByteBuf compressChunk(ByteBuf chunkBuffer, boolean isFullChunk) {
       return null;
     }
 
-    // Convert ByteBuf to byte[] before applying compression.
-    byte[] sourceBuffer;
-    int sourceBufferOffset;
-    try {
-      Triple<byte[], Integer, Integer> bufferInfo = convertByteBufToByteArray(chunkBuffer);
-      sourceBuffer = bufferInfo.getLeft();
-      sourceBufferOffset = bufferInfo.getMiddle();
-      sourceDataSize = bufferInfo.getRight();
-    } catch (Exception ex) {
-      logger.error(String.format("Cannot compress source data because failed to convert ByteBuf to byte[]."
-              + "chunkBuffer.hasArray() = %s, chunkBuffer.readableBytes = %d, chunkBuffer.arrayOffset = %d",
-          chunkBuffer.hasArray(), chunkBuffer.readableBytes(), chunkBuffer.arrayOffset()), ex);
-
-      // Emit metrics for count how many compressions skipped due to buffer conversion error.
-      compressionMetrics.compressErrorRate.mark();
-      compressionMetrics.compressErrorBufferConversion.inc();
-      return null;
-    }
-
     // Apply compression.
     CompressionMetrics.AlgorithmMetrics algorithmMetrics = compressionMetrics.getAlgorithmMetrics(defaultCompressor.getAlgorithmName());
-    Pair<Integer, byte[]> compressResult;
+    ByteBuf newChunkBuffer = combineBuffer(chunkBuffer, outputDirectBuffer, defaultCompressor.requireMatchingBufferType());
+    ByteBuf compressedBuffer = null;
+    int actualCompressedBufferSize;
     try {
       algorithmMetrics.compressRate.mark();
+
+      // Allocate compression buffer.
+      ByteBuffer sourceBuffer = newChunkBuffer.nioBuffer();
+      int compressedBufferSize = defaultCompressor.getCompressBufferSize(sourceBuffer.remaining());
+      compressedBuffer = outputDirectBuffer ? byteBufAllocator.directBuffer(compressedBufferSize) :
+          byteBufAllocator.heapBuffer(compressedBufferSize);
+
+      // Compressed buffer allocated, run compression.
       long startTime = System.nanoTime();
-      compressResult = defaultCompressor.compress(sourceBuffer, sourceBufferOffset, sourceDataSize);
+      ByteBuffer compressedByteBuffer = compressedBuffer.nioBuffer(compressedBuffer.readerIndex(), compressedBufferSize);
+      actualCompressedBufferSize = defaultCompressor.compress(sourceBuffer, compressedByteBuffer);
+      compressedBuffer.writerIndex(compressedBuffer.writerIndex() + actualCompressedBufferSize);
```

Comment:
Doesn't the compress method advance this index when writing to compressedBuf ?

Reply:
No, it does not not because compress() takes ByteBuffer, not byteBuf.